funsor/interpreter.py

```diff
@@ -216,67 +215,49 @@ def gensym(x=None):
     return "V" + str(sym)
 
 
-def stack_reinterpret(x):
-    r"""
-    Overloaded reinterpretation of a deferred expression.
-    This interpreter uses an explicit stack and no recursion but is much slower.
-
-    This handles a limited class of expressions, raising
-    ``ValueError`` in unhandled cases.
-
-    :param x: An input, typically involving deferred
-        :class:`~funsor.terms.Funsor` s.
-    :type x: A funsor or data structure holding funsors.
-    :return: A reinterpreted version of the input.
-    :raises: ValueError
-    """
-
-    def _key(x):
-        return id(x) if is_numeric_array(x) or not isinstance(x, Hashable) else x
-
-    x_name = gensym(x)
-    node_vars = {x_name: x}
-    node_names = {x: x_name}
-    env = {}
-    stack = [(x_name, x)]
-    parent_to_children = OrderedDict()
-    child_to_parents = OrderedDict()
+def anf(x):
+    stack = [x]
+    child_to_parents, children_counts = {}, {}
+    leaves = []
     while stack:
-        h_name, h = stack.pop(0)
-        parent_to_children[h_name] = []
+        h = stack.pop(0)
```

Comment:
It would be cheaper to use a [deque](https://docs.python.org/3/library/collections.html#collections.deque) for stack and leaves.